File: backend/app/schemas/national.py
from datetime import datetime

from pydantic import BaseModel, Field

from app.schemas.base import OrmSchema


class DistrictSummary(BaseModel):
    district: str
    suppressed: bool = False
    record_count: int | None = None
    low_risk: int | None = None
    moderate_risk: int | None = None
    high_risk: int | None = None
    unscored: int | None = None
    high_risk_rate: float | None = None
    latest_record_at: str | None = None


class DistrictSummaryResponse(BaseModel):
    districts: list[DistrictSummary]
    generated_at: str
    min_cell_size: int


class NationalMapMetric(BaseModel):
    id: str
    name: str
    status: str
    high_risk_share: float | None = None


class NationalMapSummaryResponse(BaseModel):
    generated_at: str
    metric_basis: str
    minimum_cell_size: int
    attribution: str
    divisions: list[NationalMapMetric]
    districts: list[NationalMapMetric]


class DivisionForecastRequest(BaseModel):
    scope_id: str = Field(min_length=1, max_length=32)


class ResourceEstimate(BaseModel):
    district: str
    suppressed: bool = False
    projected_high_risk: float | None = None
    testing_kits: int | None = None
    medicine_packs: int | None = None
    clinic_sites: int | None = None
    source: str = "burden"


class ResourceAllocationResponse(BaseModel):
    resources: list[ResourceEstimate]
    generated_at: str


class PopulationForecastJobResponse(OrmSchema):
    id: str
    status: str
    result: dict | None = None
    error_message: str | None = None
    created_at: datetime
    completed_at: datetime | None = None


class PatternInsight(BaseModel):
    statement: str
    caveat: str | None = None


class PatternDiscoveryResponse(BaseModel):
    insights: list[PatternInsight] = Field(default_factory=list)
    generated_at: str
    insufficient_data: bool = False


class GeoOption(BaseModel):
    id: str
    name: str


class GeoListResponse(BaseModel):
    items: list[GeoOption]


class IndividualPredictRequest(BaseModel):
    age: float = Field(ge=15, le=100)
    gender: str
    bmi: float = Field(gt=0, le=80)
    glucose: float = Field(gt=0, le=600)
    family_history: str
    activity: str


class IndividualPredictResponse(BaseModel):
    probability_percent: float
    risk_level: str
    category_label: str
    confidence_percent: float
    top_factors: list[str]
    engine: str = "llm"
    persisted: bool = False


class DivisionChartResponse(BaseModel):
    labels: list[str]
    high_risk_share_percent: list[float | None]
    national_high_risk_share_percent: float | None
    empty: bool


class DemographicsChartResponse(BaseModel):
    labels: list[str]
    male_high_risk_share_percent: list[float | None]
    female_high_risk_share_percent: list[float | None]
    empty: bool


class SpatialDistrictMetric(BaseModel):
    label: str
    rate: float
    severity: str
    source: str = "database"


class SpatialPanelResponse(BaseModel):
    title: str
    division_id: str
    district_id: str | None = None
    districts: list[SpatialDistrictMetric]
    metrics: dict
    metric_basis: str

File: frontend/lib/api/national.ts
import { api } from './client'

export type GeoOption = { id: string; name: string }

export type DistrictSummary = {
  district: string
  suppressed: boolean
  record_count: number | null
  low_risk: number | null
  moderate_risk: number | null
  high_risk: number | null
  unscored: number | null
  high_risk_rate: number | null
  latest_record_at: string | null
}

export type DistrictSummaryResponse = {
  districts: DistrictSummary[]
  generated_at: string
  min_cell_size: number
}

export type NationalMapMetric = {
  id: string
  name: string
  status: 'available' | 'awaiting_scores' | 'suppressed' | 'unavailable'
  high_risk_share: number | null
}

export type NationalMapSummary = {
  generated_at: string
  metric_basis: string
  minimum_cell_size: number
  attribution: string
  divisions: NationalMapMetric[]
  districts: NationalMapMetric[]
}

export type ResourceEstimate = {
  district: string
  suppressed: boolean
  projected_high_risk: number | null
  testing_kits: number | null
  medicine_packs: number | null
  clinic_sites: number | null
  source: string
}

export type ResourceAllocationResponse = {
  resources: ResourceEstimate[]
  generated_at: string
}

export type PopulationForecastJob = {
  id: string
  status: string
  result: {
    districts?: Array<{
      district: string
      status: string
      model: string
      high_risk_6m: number | null
      high_risk_12m: number | null
      record_count_6m: number | null
      record_count_12m: number | null
    }>
  } | null
  error_message: string | null
  created_at: string
  completed_at: string | null
}

export type PatternInsight = {
  statement: string
  caveat: string | null
}

export type PatternDiscoveryResponse = {
  insights: PatternInsight[]
  generated_at: string
  insufficient_data: boolean
}

export type SpatialPanel = {
  title: string
  division_id: string
  district_id: string | null
  districts: Array<{ label: string; rate: number; severity: string; source: 'database' }>
  metrics: {
    high_risk_share_change_yoy_percentage_points: number | null
    mean_record_age_years: number | null
    screening_coverage_percent: number | null
  }
  metric_basis: string
}

export type DivisionChart = {
  labels: string[]
  high_risk_share_percent: Array<number | null>
  national_high_risk_share_percent: number | null
  empty: boolean
}

export type DemographicsChart = {
  labels: string[]
  male_high_risk_share_percent: Array<number | null>
  female_high_risk_share_percent: Array<number | null>
  empty: boolean
}

export type IndividualPredictRequest = {
  age: number
  gender: string
  bmi: number
  glucose: number
  family_history: string
  activity: string
}

export type IndividualPredictResponse = {
  probability_percent: number
  risk_level: string
  category_label: string
  confidence_percent: number
  top_factors: string[]
  engine: string
  persisted: boolean
}

export type EpidemicForecastJob = {
  id: string
  status: string
  result: {
    forecast_kind?: string
    pending?: boolean
    engine?: string
    summary?: string
    series?: Array<{
      year: number
      urban_prevalence_percent: number
      rural_prevalence_percent: number
    }>
    risk_groups?: Array<{
      key: string
      label: string
      population_share_percent: number
      guidance: string
    }>
  } | null
  error_message: string | null
  created_at: string
  completed_at: string | null
}

export type DivisionForecastPoint = {
  year: number
  value: number | null
  kind: 'observed' | 'projected' | 'unavailable'
  source: string | null
}

export type DivisionForecastJob = {
  id: string
  status: string
  result: {
    forecast_kind?: string
    scope_id?: string
    engine?: string
    metric_basis?: string
    minimum_cell_size?: number
    display_years?: number[]
    projection_start_year?: number | null
    urban_rural_status?: string
    summary?: string
    series?: Array<{
      id: string
      name: string
      status: string
      urban_rural_status: string
      points: DivisionForecastPoint[]
    }>
  } | null
  error_message: string | null
  created_at: string
  completed_at: string | null
}

const ACCESS_TOKEN_KEY = 'health_access_token'

async function downloadCsv(): Promise<{ error?: string }> {
  const token = typeof window !== 'undefined' ? localStorage.getItem(ACCESS_TOKEN_KEY) : null
  try {
    const res = await fetch('/v1/national/export.csv', {
      headers: token ? { Authorization: `Bearer ${token}` } : {},
    })
    if (!res.ok) {
      const body = await res.json().catch(() => ({}))
      return { error: (body as { detail?: string }).detail || `Export failed (${res.status})` }
    }
    const blob = await res.blob()
    const url = URL.createObjectURL(blob)
    const a = document.createElement('a')
    a.href = url
    a.download = 'national-district-summary.csv'
    a.click()
    URL.revokeObjectURL(url)
    return {}
  } catch (e) {
    return { error: e instanceof Error ? e.message : 'Network error' }
  }
}

export const nationalApi = {
  getDivisions: () => api.get<{ items: GeoOption[] }>('/national/geo/divisions'),
  getDistricts: (divisionId: string) =>
    api.get<{ items: GeoOption[] }>(`/national/geo/districts?division_id=${encodeURIComponent(divisionId)}`),
  getSpatial: (divisionId: string, districtId?: string) => {
    const q = new URLSearchParams({ division_id: divisionId })
    if (districtId) q.set('district_id', districtId)
    return api.get<SpatialPanel>(`/national/spatial?${q}`)
  },
  getDivisionChart: () => api.get<DivisionChart>('/national/charts/divisions'),
  getDemographicsChart: () => api.get<DemographicsChart>('/national/charts/demographics'),
  predictIndividual: (body: IndividualPredictRequest) =>
    api.post<IndividualPredictResponse>('/national/predict-individual', body),
  enqueueEpidemicForecast: () => api.post<EpidemicForecastJob>('/national/epidemic-forecast'),
  getLatestEpidemicForecast: () => api.get<EpidemicForecastJob>('/national/epidemic-forecast/latest'),
  enqueueDivisionForecast: (scopeId: string) => api.post<DivisionForecastJob>('/national/division-forecasts', { scope_id: scopeId }),
  getLatestDivisionForecast: (scopeId: string) =>
    api.get<DivisionForecastJob>(`/national/division-forecasts/latest?scope_id=${encodeURIComponent(scopeId)}`),
  getDistrictSummary: () => api.get<DistrictSummaryResponse>('/national/districts/summary'),
  getMapSummary: () => api.get<NationalMapSummary>('/national/map/summary'),
  getResources: () => api.get<ResourceAllocationResponse>('/national/resources'),
  exportCsv: downloadCsv,
  enqueueForecast: () => api.post<PopulationForecastJob>('/national/forecasts'),
  getLatestForecast: () => api.get<PopulationForecastJob>('/national/forecasts/latest'),
  discoverPatterns: () => api.post<PatternDiscoveryResponse>('/national/patterns'),
}
